src/test-run/index.js

```diff
@@ -261,3 +300,42 @@ ServiceMessages[CLIENT_MESSAGES.ready] = function (msg) {
 
     return new Promise((resolve, reject) => this.pendingRequest = { resolve, reject, responseTimeout });
 };
+
+ServiceMessages[CLIENT_MESSAGES.executePendingBrowserManipulation] = async function (msg) {
+    this.debugLog.driverMessage(msg);
+
+    var command        = this.pendingWindowManipulation;
+    var executionError = null;
+
+    if (command.type === COMMAND_TYPE.takeScreenshot) {
+        try {
+            await this.screenshotCapturer.captureAction({
+                pageUrl:    msg.pageUrl,
+                stepName:   uuid.v4(),
+                customPath: command.customPath
+            });
+        }
+        catch (err) {
+            // NOTE: swallow the error silently if we can't take screenshots for some
+            // reason (e.g. we don't have permissions to write a screenshot file).
+            executionError = null;
+        }
+    }
+    else if (command.type === COMMAND_TYPE.takeScreenshotOnFail) {
+        try {
+            this.screenshotPath = await this.screenshotCapturer.captureError({
+                pageUrl:            msg.pageUrl,
+                stepName:           uuid.v4(),
+                screenshotRequired: this.opts.takeScreenshotsOnFails
+            });
+        }
+        catch (err) {
+            // NOTE: swallow the error silently if we can't take screenshots for some
+            // reason (e.g. we don't have permissions to write a screenshot file).
+            executionError      = null;
+            this.screenshotPath = SCREENSHOT_CAN_NOT_BE_SAVED_MESSAGE;
```

Comment:
I don't like that we use state variable, because it creates action side effect. Can we return `screenshotPath` as result of `TakeScreenshotOnFailCommand` action?